src/input/joystick.c

```diff
@@ -0,0 +1,714 @@
+    NUM_DIRECTIONS = 4
+};
+
+typedef enum {
+    INPUT_STATE_IS_UP,
+    INPUT_STATE_WENT_DOWN,
+    INPUT_STATE_IS_DOWN,
+    INPUT_STATE_WENT_UP
+} input_state;
+
+typedef struct {
+    joystick_button top;
+    joystick_button left;
+    joystick_button bottom;
+    joystick_button right;
+} joystick_hat;
+
+typedef struct {
+    int delta_x;
+    int delta_y;
+} joystick_trackball;
+
+typedef struct {
+    int id;
+    int connected;
+    joystick_model *model;
+    joystick_axis axis[JOYSTICK_MAX_AXIS];
+    joystick_button button[JOYSTICK_MAX_BUTTONS];
+    joystick_trackball trackball[JOYSTICK_MAX_TRACKBALLS];
+    joystick_hat hat[JOYSTICK_MAX_HATS];
+} joystick_info;
+
+typedef struct {
+    joystick_element element;
+    int value;
+    input_state state;
+} mapped_input;
+
+enum {
+    CURSOR_SLOWDOWN_FASTER = 1024,
+    CURSOR_SLOWDOWN_NORMAL = 4096,
+    CURSOR_SLOWDOWN_SLOWER = 8192
+};
+
+#define JOYSTICK_HOTKEY_OFFSET MAPPING_ACTION_ROTATE_MAP_LEFT
+#define JOYSTICK_MAX_HOTKEYS 6
+
+static struct {
+    joystick_model connected_models[JOYSTICK_MAX_CONTROLLERS];
+    joystick_info joystick[JOYSTICK_MAX_CONTROLLERS];
+    int connected_joysticks;
+    struct {
+        speed_type x_speed;
+        speed_type y_speed;
+        int left_button;
+        int middle_button;
+        int right_button;
+        scroll_state scroll;
+        time_millis last_scroll_time;
+    } mouse;
+    mapped_input map_scroll[NUM_DIRECTIONS];
+    mapped_input joystick_hotkey[JOYSTICK_MAX_HOTKEYS];
+    mapped_input virtual_keyboard;
+    mapped_input touch_mode;
+} data;
+
+static hotkey_action JOYSTICK_MAPPING_TO_HOTKEY_ACTION[JOYSTICK_MAX_HOTKEYS] = {
+    HOTKEY_ROTATE_MAP_LEFT,
+    HOTKEY_ROTATE_MAP_RIGHT,
+    HOTKEY_INCREASE_GAME_SPEED,
+    HOTKEY_DECREASE_GAME_SPEED,
+    HOTKEY_TOGGLE_PAUSE,
+    HOTKEY_CYCLE_LEGION
+};
+
+joystick_model *joystick_get_model_by_guid(const char *guid)
+{
+    for (int i = 0; i < JOYSTICK_MAX_CONTROLLERS; ++i) {
+        joystick_model *model = &data.connected_models[i];
+        if (strcmp(guid, model->guid) == 0) {
+            return model;
+        }
+    }
+    return 0;
+}
+
+joystick_model *joystick_get_first_unused_model(void)
+{
+    for (int i = 0; i < JOYSTICK_MAX_CONTROLLERS; ++i) {
+        joystick_model *model = &data.connected_models[i];
+        if (!model->connected_joysticks) {
+            return model;
+        }
+    }
+    return 0;
+}
+
+static void update_hat(joystick_hat *hat, joystick_hat_position position)
+{
+    hat->top = (position & JOYSTICK_HAT_UP) ? 1 : 0;
+    hat->left = (position & JOYSTICK_HAT_LEFT) ? 1 : 0;
+    hat->bottom = (position & JOYSTICK_HAT_DOWN) ? 1 : 0;
+    hat->right = (position & JOYSTICK_HAT_RIGHT) ? 1 : 0;
+}
+
+static void update_trackball(joystick_trackball *trackball, int delta_x, int delta_y)
+{
+    trackball->delta_x += delta_x;
+    trackball->delta_y += delta_y;
+}
+
+static void joystick_reset_all_button_states(joystick_info *joystick)
```

Comment:
Preferably don't prefix `static` functions with the name of the file (`joystick_`): that way it's easier to see from the name whether the function is public or not.
Also, I think these one-liner functions can be inlined in `joystick_reset_state`.